### PR TITLE
Generalize passes that require RTTI

### PR DESCRIPTION
**Pull request by peterschrammel**
A small extension to languaget allows us to make the following passes language-independent:
* remove_virtual_methods
* remove_exceptions
* remove_instanceof

The caveat is that (ignoring CPP which needs to be reimplemented) there is currently no other language that would use these facilities.

Alternatively, we could choose to move these passes to java_bytecode, with the (at the moment not much worrying) risk that incompatible infrastructures for similar functionalities will be re-implemented for future languages.

### CODE REVIEW

**Comment by smowton** [comment]
Seems sensible -- two possible changes:
* Instantiating a `languaget` to answer the question "what is your class hierarchy root?", and doing that once per `instanceof` expression, seems rather heavyweight. How about introducing some `language_infot` or similar, which is stateless, and which is instantiated once and cached to answer questions like this?
* Why are you removing the `remove_exceptions` special case for `AssertionError`?

File: src/goto-analyzer/goto_analyzer_parse_options.cpp
Comment (tautschnig) [inline src/goto-analyzer/goto_analyzer_parse_options.cpp]: Nit pick: lexicographic ordering

File: src/goto-diff/goto_diff_parse_options.cpp
Comment (tautschnig) [inline src/goto-diff/goto_diff_parse_options.cpp]: Nit pick: lexicographic ordering

File: src/jbmc/jbmc_parse_options.cpp
Comment (tautschnig) [inline src/jbmc/jbmc_parse_options.cpp]: Nit pick: lexicographic ordering

File: src/goto-programs/remove_exceptions.cpp
Comment (tautschnig) [inline src/goto-programs/remove_exceptions.cpp]: Should there be a test to cover this?

**Review by tautschnig — APPROVED** [review]
Looks good to me, but I can't claim serious java_bytecode knowledge.

**Comment by peterschrammel** [comment]
@smowton, yes, that's indeed very, very inefficient. We cannot transparently cache the language in the global languages list, but we could provide a caching wrapper that could be used where necessary.

**Comment by smowton** [comment]
@peterschrammel in fact on closer inspection we sort of have this mechanism already -- the `languge_entryt` used in `langapi/mode.cpp` does something very similar, avoiding needing to instantiate `languaget` in order to answer "what file extensions are associated with this language," or "what mode is associated with this language?" How about extending `language_entryt`, or having registering languages supply a `language_entryt` instead of just a factory?

**Comment by thk123** [comment]
Am holding off on review this due to above comment suggests a potential re-work - let me know if I should review as is.

**Comment by martin-cs** [comment]
+1, what @thk123 said.  Am happy to review if this is live again.

FWIW Ada will need something like remove virtual functions and remove exceptions and I know that @andreast271 is working on C++ and so might care about having space for that.

**Comment by peterschrammel** [comment]
I'm going to revive this. 
It needs a bit of rework since we are moving from a 'framework-provides-dispatch' to a 'driver-provides-dispatch' mechanism for language-specific features.

**Comment by thomasspriggs** [comment]
This needs rebasing due to a number of merge conflicts and running out of date CI jobs. @peterschrammel are you still planning to revive this? If this is work which is no longer planned then we should close it out in order to clean up the number of open PRs.

**Comment by peterschrammel** [comment]
It's probably easier to reimplement these few lines than to rebase them.

**Pull request closed without merging**
